lib/Auth/Process/ExpiryDate.php

```diff
@@ -0,0 +1,356 @@
+            $relayState = $state['saml:RelayState'];
+            if (strpos($relayState, $changePwdUrl) !==false) {                
+                SimpleSAML_Auth_ProcessingChain::resumeProcessing($state);
+        // otherwise add the original destination url as a parameter
+            } else {
+                $returnTo = sspmod_expirychecker_Utilities::getUrlFromRelayState(
+                                                                        $relayState);
+                if ($returnTo) {                                 
+                    $changePwdUrl = $changePwdUrl . "?returnTo=" . $returnTo;
+                }
+            }
+        }
+
+        $this->logger->warning(sprintf(
+            'expirychecker: Password for %s is about to expire, redirecting to %s',
+            var_export($accountName, true),
+            var_export($changePwdUrl, true)
+        ));
+
+        SimpleSAML_Utilities::redirect($changePwdUrl, array());
+    }
+    
+    /**
+     * Apply this AuthProc Filter.
+     *
+     * @param array &$state The current state.
+     */
+    public function process(&$state)
+    {
+        // Get the necessary info from the state data.
+        $accountName = $this->getAttribute($this->accountNameAttr, $state);
+        $expiryTimestamp = $this->getExpiryTimestamp($this->expiryDateAttr, $state);
+        
+        if ($this->isExpired($expiryTimestamp)) {
+            $this->redirectToExpiredPage($state, $accountName, $expiryTimestamp);
+        }
+        
+        // Display a password expiration warning page if it's time to do so.
+        if ($this->isTimeToWarn($expiryTimestamp, $this->warnDaysBefore)) {
+            $this->redirectToWarningPage($state, $accountName, $expiryTimestamp);
+        }
+    }
+    
+    /**
+     * Redirect the user to the expired-password page.
+     *
+     * @param array $state The state data.
+     * @param string $accountName The name of the user account.
+     * @param int $expiryTimestamp When the password expired.
+     */
+    public function redirectToExpiredPage(&$state, $accountName, $expiryTimestamp)
+    {
+        assert('is_array($state)');
+        
+        $expireOnDate = date($this->dateFormat, $expiryTimestamp);
+        
+        $this->logger->error(sprintf(
+            'expirychecker: Password for %s has expired [%s]. Access denied.',
+            var_export($accountName, true),
+            $expireOnDate
+        ));
+
+        /* Save state and redirect. */
+        $state['expireOnDate'] = $expireOnDate;
+        $state['accountName'] = $accountName;
+        $state['changePwdUrl'] = $this->changePwdUrl;
+        $state['originalUrlParam'] = $this->originalUrlParam;
+        
+        $id = SimpleSAML_Auth_State::saveState($state, 'expirychecker:expired');
+        $url = SimpleSAML_Module::getModuleURL('expirychecker/expired.php');
+        
+        SimpleSAML_Utilities::redirect($url, array('StateId' => $id));
```

Comment:
Is that `StateId` parameter anything like a session id that should not be in the url?